### PR TITLE
rework yum-requirements

### PR DESCRIPTION
**Pull request by mariusvniekerk**
This removes the need to specially deal with yum_requirements in conda-smithy itself.  This allows us to get rid of a fair bit of code and remove the need to rerender when just changing yum_requirements

### CODE REVIEW

File: conda_smithy/templates/run_docker_build.sh.tmpl [hunk 1]
```diff
@@ -39,6 +39,27 @@ mkdir -p "$ARTIFACTS"
...
+pushd ${RECIPE_ROOT}
+if [ -f ${RECIPE_ROOT}/yum_requirements.txt ]; then
+
+    cat >Dockerfile.yum <<DOCKERFILE
+FROM ${DOCKER_IMAGE}
+ADD yum_requirements.txt /opt/requirements.txt
+RUN yum -y install \$(cat /opt/requirements.txt)
+DOCKERFILE
+
+    export NEW_DOCKER_IMAGE="conda-smithy-$(openssl rand -hex 12)"
```
Comment (isuruf) [inline conda_smithy/templates/run_docker_build.sh.tmpl]: Can the dependency on openssl be dropped? I'd like to have `run_docker_build.sh` working without additional dependencies.

Comment (mariusvniekerk) [reply conda_smithy/templates/run_docker_build.sh.tmpl]: since this is used with git, you will have openssl around?

Comment (mariusvniekerk) [reply conda_smithy/templates/run_docker_build.sh.tmpl]: but we can change this to antyhing that emits a randomish number, this was just the easiest way

**Comment by mariusvniekerk** [comment]
This also allows us to remove sudo from our docker images.

File: conda_smithy/templates/run_docker_build.sh.tmpl (hunk 1 above)
Comment (isuruf) [reply conda_smithy/templates/run_docker_build.sh.tmpl]: Why not do this inside the docker container? There you can have openssl installed via conda

**Comment by isuruf** [comment]
> This also allows us to remove sudo from our docker images.

I'm -1 to that change. This makes using the docker image for debugging harder for no useful benefit. What's the reason to remove sudo?

**Comment by mariusvniekerk** [comment]
In our current builds we cannot run sudo at all for aarch64 or ppc64le.  This means we cannot yum install anything whatsoever.  There are two viable solution mechanisms to this problem, either check for the existence of the yum_requirements at the docker entrypoint prior to switching to the conda user, or building a new image as in this PR

File: conda_smithy/templates/run_docker_build.sh.tmpl (hunk 1 above)
Comment (mariusvniekerk) [reply conda_smithy/templates/run_docker_build.sh.tmpl]: we need the name of the docker image in order to run it....

**Comment by mariusvniekerk** [comment]
We don't really need to remove sudo. but we also don't need to use it anymore either

**Comment by isuruf** [comment]
> In our current builds we cannot run sudo at all for aarch64 or ppc64le.

Why can we do this for the x86_64 build?

**Comment by mariusvniekerk** [comment]
The core issue is mostly around native vs emulated execution.  On native systems it seems that sudo works as expected, but running sudo under qemu when you are not uid 0 just fails with messages about not being uid 0.   In the interest of keeping the build approaches mostly the same we change both.

**Comment by mariusvniekerk** [comment]
Example PR where this was applied.  https://github.com/conda-forge/tk-feedstock/pull/32

**Comment by isuruf** [comment]
Does https://github.com/multiarch/qemu-user-static/issues/17#issuecomment-434423150 work?

**Comment by mariusvniekerk** [comment]
how on earth did you find that.....,  dunno, it might

File: conda_smithy/templates/run_docker_build.sh.tmpl (hunk 1 above)
Comment (jakirkham) [reply conda_smithy/templates/run_docker_build.sh.tmpl]: Sorry for being dense, why do we need a random number here?

**Comment by mariusvniekerk** [comment]
looks like that does work.  See https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=13649

do we want to keep some of the bits around how we execute yum so that adding yum requirements does not require a rerender?

**Comment by isuruf** [comment]
I'm okay with keeping this if it was moved inside the docker container

**Comment by mariusvniekerk** [comment]
@isuruf made that change

**Comment by mariusvniekerk** [comment]
cc @hmaarrfk @jjhelmus

**Comment by hmaarrfk** [comment]
whoa, magical!!!

File: conda_smithy/templates/build_steps.sh.tmpl [hunk 2]
```diff
@@ -26,8 +26,13 @@ setup_conda_rc "${FEEDSTOCK_ROOT}" "${RECIPE_ROOT}" "${CONFIG_FILE}"
 
 {% if build_setup -%}
 {{ build_setup }}{% endif -%}
-{% if yum_build_setup is defined -%}
-{{ yum_build_setup }}{% endif -%}
+
+# If we have yum requirements build a new specialized image that we will use to run our build in that includes
+# those dependencies.
```
Comment (isuruf) [inline conda_smithy/templates/build_steps.sh.tmpl]: Can you update this comment or remove it?

Comment (jakirkham) [reply conda_smithy/templates/build_steps.sh.tmpl]: Actually are we able to just revert all of the changes to this file now?

File: conda_smithy/templates/run_docker_build.sh.tmpl [hunk 3]
```diff
@@ -53,15 +56,17 @@ DOCKER_RUN_ARGS=" -it "
...
-           /home/conda/feedstock_root/${PROVIDER_DIR}/build_steps.sh
+    -v "${RECIPE_ROOT}":/home/conda/recipe_root:ro,z \
+    -v "${FEEDSTOCK_ROOT}":/home/conda/feedstock_root:rw,z \
+    -e CONFIG \
+    -e BINSTAR_TOKEN \
+    -e HOST_USER_ID \
+    -e UPLOAD_PACKAGES \
+    $DOCKER_IMAGE \
+    {{ docker.command }} \
+    /home/conda/feedstock_root/${PROVIDER_DIR}/build_steps.sh
```
Comment (jakirkham) [inline conda_smithy/templates/run_docker_build.sh.tmpl]: Also are we able to revert the changes here too?

**Comment by jakirkham** [comment]
Since @isuruf's suggestion worked, would encourage slimming this down to the minimal required change. If I'm not mistaken, that amounts to one line. Though please feel free to correct me if I'm wrong.

File: conda_smithy/templates/run_docker_build.sh.tmpl (hunk 1 above)
Comment (jakirkham) [reply conda_smithy/templates/run_docker_build.sh.tmpl]: FWIW I also share @isuruf's concerns with increasing reliance on what is outside the Docker image. Raised issue ( https://github.com/conda-forge/conda-smithy/issues/1020 ) to discuss cutting back on existing things that are required outside the image to make builds easier to reproduce and debug locally.

File: conda_smithy/templates/run_docker_build.sh.tmpl (hunk 3 above)
Comment (mariusvniekerk) [reply conda_smithy/templates/run_docker_build.sh.tmpl]: we could, but this is just whitespace

File: conda_smithy/templates/build_steps.sh.tmpl (hunk 2 above)
Comment (mariusvniekerk) [reply conda_smithy/templates/build_steps.sh.tmpl]: we could, though this does add the advantage that we dont need a rerender if we've changed yum_requirements

File: conda_smithy/templates/run_docker_build.sh.tmpl (hunk 3 above)
Comment (jakirkham) [reply conda_smithy/templates/run_docker_build.sh.tmpl]: Exactly

File: conda_smithy/templates/build_steps.sh.tmpl (hunk 2 above)
Comment (jakirkham) [reply conda_smithy/templates/build_steps.sh.tmpl]: Yeah, this is roughly how I proposed `yum_requirements.txt` originally ( https://github.com/conda-forge/conda-smithy/pull/135 ), but Phil made a compelling argument that they should be listed explicitly ( https://github.com/conda-forge/conda-smithy/pull/134 ). Ultimately I don't feel strongly either way, but it seems the crux of the needed change is quite simple; so, it would be nice if this PR reflected that. Orthogonal changes are better left for other PRs IMHO.

**Comment by scopatz** [comment]
Wow, that sudo change is insane! Great find @isuruf!
